### DISCUSSION TITLE
How to invoke a ReactiveCommand and wait for it to finish?

### DISCUSSION THREAD

**Discussion by LJN-hzleaper** [reactiveui/ReactiveUI#3615] [Q&A] [answered]
If a ReactiveCommand body is a async method and has a await statement, `cmd.Execute().Subscribe();` will return in advance when program run to the await statement, how to wait for the whole command body to finish?
For example, I create a ReactiveCommand by this:
```
CmdQueryDoffIdList = ReactiveCommand.CreateFromObservable(
            () => Observable.StartAsync(QueryDoffIdList));

private async Task QueryDoffIdList(CancellationToken token){
    // Do something A,,, 
   await Model.QueryDoffIdListAsync(VmQueryCondition, token);
   // Do something B...
}
```
and in somewhere I execute the command by this:
```
CmdQueryDoffIdList.Execute().Subscribe();
```
The `Subscribe()` method call will return when program run to the `await Model.QueryDoffIdListAsync(VmQueryCondition, token);` line , but I want it return after "Do something B...". How to implement this?

**Comment by LJN-hzleaper** [comment, marked as answer]
I found the solution, I can use "await CmdQueryDoffIdList.Execute().ToTask();"

**Reply by ChrisPulman** [reply]
You should be able to await any IObservable<T> and get the first result back without the need for the ToTask(), just bear in mind that if you are expecting a stream of values you will only get the first, ReactiveCommand only gives one Result per execution so will be okay. 

In the Observer there is OnNext, OnError and Complete that may be more appropriate but are non blocking like await so a different approach would be required if your wanting to have a blocked section of code; such as a SemaphoreLite which you set prior to calling the command and release on Complete then await it after your subscription, you can then collect all the results from your Observer and the Semaphore will be released on Completion giving you all the results. There's many ways to do this and it depends on your personal requirements for the functionality you need as to how you achieve it.

**Reply by LJN-hzleaper** [reply]
Thanks
